### ISSUE TITLE
Infinite loop after `break` mutation

### ISSUE THREAD

**Issue by nepalez** [mbj/mutant#443]
Hi, Markus!

Is there any possibility to prevent mutation (exclusion) of `break` statement?

Seems like mutant does it in the code below, and then falls into infinite loop:

```ruby
# Iterates through Kafka messages
#
# Fetches the next portion of messages until the result is empty
#
# @return [Enumerator]
#
# @yieldparam [Hash{Symbol => String, Integer}] tuple
#
def each
  return to_enum unless block_given?
  loop do
    tuples = fetch
    tuples.any? ? tuples.each { |tuple| yield(tuple) } : break
  end
end
```

**Comment by nepalez** [comment]
Hmm, maybe this is a problem of my ruby settings :(

**Comment by mbj** [comment]
> Is there any possibility to prevent mutation (exclusion) of break statement?

Intentionally not.

> Seems like mutant does it in the code below, and then falls into infinite loop:

Correct. Under the assuption `fetch` does always return new tuples.

But there are multiple ways to kill this mutation, in order of preference descending:

* Assert `#fetch` does only get call a certain amount of times (maybe inject the data source)
* Set a timeout to kill the infinite mutations from specs
* Ignore this specific subject

I'm happy to show you how to apply especially the first one. In case you provide me with a minimum reproduction that runs into the inifinte loop (or simple the link to the public repo), I could PR the fix for the mutation.

**Comment by nepalez** [comment]
Thank you for the proposal!

The repository is [here](https://github.com/nepalez/rom-kafka).
This is [the line of code](https://github.com/nepalez/rom-kafka/blob/master/lib/rom/kafka/connection/consumer.rb#L92), where 2 looped mutation occurs. And [here is the coverage](https://github.com/nepalez/rom-kafka/blob/master/spec/unit/connection/consumer_spec.rb#L77-L78).

When I manually kill those 2 evil processes, mutant gracefully reports there is no unkilled evils.

**Comment by nepalez** [comment]
After reading [this old thread](https://github.com/mbj/mutant/issues/265), I set timeout to `0.5` and all mutations were killed (under MRI 2.1.5).

**Comment by mbj** [comment]
@nepalez Another option, and IMO one to prefer before going into time based killing: Is to provide an object that raises when `#fetch` is called more than 3 times.

Message expectation counts by rspec only get asserted AFTER the code under test finished.

If you are still interested in me killing the mutation say so. Or can we close this issue?

**Comment by nepalez** [comment]
Ahh, thanks, this option is interesting as well. I'll try it.

**Comment by mbj** [comment]
k, cool.

Rule of thump: look for observable behavior changes by a mutation. In a highly mutable language like ruby there are plenty.

**Comment by nepalez** [comment]
Well, I change it to raise an exception when excessive step is done. (And removed timeout, because you're right - it is sort of magics, too context-dependent).

It doesn't solved the case totally. 1 mutation still cause an infinite loop. I was wondered and look at it. This mutation simply deletes ALL the content of the loop block :) and left this chunk of code:

```ruby
loop do
end
```

It seems there is no way to deal with this except for via segfault. I rewritten the method from `loop - break` to `do - while`. Now this works (though rubocop grumbles about this substitution) and mutant reports 100% coverage.

To recap, this sort of mutation seems buggy in the concrete case.

**Comment by mbj** [comment]
```
loop do
end
```
That one can only be killed via the timeout. I think its easy to filter it out. Since a content less infinite loop almost never makes sense, and its very uneasy to catch by tests.

Do you mind to create an issue for guarding against that one? Its a 1loc change to mutant, but I'll probably forget to do it unless there is an issue.

> `do - while` (though rubocop grumbles about this substitution)

Postcontrol loops are actually not that bad IMO. Using an infinite loop and explicit control via `break` is IMO more powerful than a loop that is *directly* controlled. So I do not necessarily agree with rubocop and on network code I write I often white-list this construct for the main "event loop".

**Comment by mbj** [comment]
@nepalez Since recent mutant releases you can timeout loops to infinite mutations mutant side, and also count them as kills:https://github.com/mbj/mutant/blob/master/docs/configuration.md#mutation_timeout
